src/condor_utils/dc_coroutines.cpp

```diff
@@ -0,0 +1,103 @@
+#include "condor_common.h"
+#include "condor_daemon_core.h"
+#include "dc_coroutines.h"
+
+
+dc::AwaitableDeadlineReaper::AwaitableDeadlineReaper() {
+	reaperID = daemonCore->Register_Reaper(
+		"AwaitableDeadlineReaper::reaper",
+		(ReaperHandlercpp) & AwaitableDeadlineReaper::reaper,
+		"AwaitableDeadlineReaper::reaper",
+		this
+	);
+}
+
+
+dc::AwaitableDeadlineReaper::~AwaitableDeadlineReaper() {
+	// Do NOT destroy() the_coroutine here.  The coroutine may still
+	// needs it state, because the lifetime of this object could be
+	// shorter than the lifetime of the coroutine.  (For example, a
+	// coroutine could declare two locals of this type, one of which
+	// has a longer lifetime than the other.)
+
+	// Cancel the reaper.  (Which holds a pointer to this.)
+	if( reaperID != -1 ) {
+		daemonCore->Cancel_Reaper( reaperID );
+	}
+
+	// Cancel any timers.  (Each holds a pointer to this.)
+	for( auto [timerID, dummy] : timerIDToPIDMap ) {
+		daemonCore->Cancel_Timer( timerID );
+	}
+}
+
+
+bool
+dc::AwaitableDeadlineReaper::born( int pid, int timeout ) {
+	auto [dummy, inserted] = pids.insert(pid);
+	if(! inserted) { return false; }
+	dprintf( D_ZKM, "Inserted %d into  %p\n", pid, & pids );
+
+	// Register a timer for this process.
+	int timerID = daemonCore->Register_Timer(
+		timeout, TIMER_NEVER,
+		(TimerHandlercpp) & AwaitableDeadlineReaper::timer,
+		"AwaitableDeadlineReaper::timer",
+		this
+	);
+	timerIDToPIDMap[timerID] = pid;
+
+	return true;
+}
+
+
+int
+dc::AwaitableDeadlineReaper::reaper_id() {
```

Comment:
Should this method be declared "const"?

Reply:
Yes, and it should be moved into the header, as well.  Left-over from it used to actually do something (as opposed to RAII).